### PR TITLE
[TOPI] Bitserial dense operators for CPU

### PR DESCRIPTION
**Pull request by cowanmeg**
TOPI cpu bitserial dense operators for low bit-widths. These very similar to the bitserial conv2d operators.

Please review @vinx13 @eqy @ajtulloch 
Thanks!

### CODE REVIEW

File: topi/python/topi/arm_cpu/bitserial_dense.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+from __future__ import absolute_import as _abs
+import tvm
+from tvm import autotvm
+from topi.util import get_const_tuple
+from .. import tag
+from .. import generic
+from .bitserial_conv2d import _intrin_popcount
+from ..nn.pad import pad
+from ..nn.bitserial_dense import bitserial_dense
+from ..nn.bitserial_conv2d import bitpack # Pull out into a utility function?
```
Comment (vinx13) [inline topi/python/topi/arm_cpu/bitserial_dense.py]: +1 for moving to a utility function

File: topi/python/topi/arm_cpu/bitserial_dense.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        fused = s[output].fuse(xo, yo)
+        s[output].parallel(fused)
+
+        nfactor = cfg['tile_y'].size[-1]
+        kfactor = cfg['tile_k'].size[-1]
+        if nfactor % 8 == 0:
+            pc = _intrin_popcount(nfactor, kfactor, WB, DB, unipolar)
+            s[output].tensorize(wb, pc)
+
+        s = s.normalize()
```
Comment (vinx13) [inline topi/python/topi/arm_cpu/bitserial_dense.py]: is it necessary here? `normalize` is called in build_module before `InferBound`

Comment (cowanmeg) [reply topi/python/topi/arm_cpu/bitserial_dense.py]: good catch, thanks!

File: topi/python/topi/arm_cpu/bitserial_dense.py (hunk 1 above)
Comment (cowanmeg) [reply topi/python/topi/arm_cpu/bitserial_dense.py]: moved to bitserial_util.py

File: topi/python/topi/arm_cpu/bitserial_dense.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    if pack_dtype == 'uint8':
+        binary_op_multiplier = 8
+    elif pack_dtype == 'uint16':
+        binary_op_multiplier = 16
+    elif pack_dtype == 'uint32':
+        binary_op_multiplier = 32
+    elif pack_dtype == 'uint64':
+        binary_op_multiplier = 64
+
+    cfg.add_flop(batch * out_dim * in_dim * binary_op_multiplier)
```
Comment (eqy) [inline topi/python/topi/arm_cpu/bitserial_dense.py]: Does this match how we usually do flop calculation (e.g., including ops for padding) etc.?

File: topi/python/topi/arm_cpu/bitserial_dense.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        << (wb+db).astype(out_dtype), axis=[wb, db, k]), tag='bitserial_dense')
+
+    if pack_dtype == 'uint8':
+        binary_op_multiplier = 8
+    elif pack_dtype == 'uint16':
+        binary_op_multiplier = 16
+    elif pack_dtype == 'uint32':
+        binary_op_multiplier = 32
+    elif pack_dtype == 'uint64':
+        binary_op_multiplier = 64
```
Comment (eqy) [inline topi/python/topi/arm_cpu/bitserial_dense.py]: nit: can we pattern match on the dtype str?

File: topi/python/topi/nn/bitserial_dense.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        << (db+wb).astype(out_dtype), axis=[wb, db, k]), tag='bitserial_dense')
+
+    if pack_dtype == 'uint8':
+        binary_op_multiplier = 8
+    elif pack_dtype == 'uint16':
+        binary_op_multiplier = 16
+    elif pack_dtype == 'uint32':
+        binary_op_multiplier = 32
+    elif pack_dtype == 'uint64':
+        binary_op_multiplier = 64
```
Comment (eqy) [inline topi/python/topi/nn/bitserial_dense.py]: "

File: topi/python/topi/nn/bitserial_util.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+def bitpack(data, bits, pack_axis, bit_axis, pack_type, name="QuantizeInput"):
+    """Packs data into format necessary for bitserial computation
+    pack_axis : int
+       index of the axis to pack in data
+    bit_axis : int
+       index of axis to place bit axis in resulting packed data"""
+    ishape = data.shape
+    n = len(ishape)
+    if pack_type == 'uint8':
```
Comment (eqy) [inline topi/python/topi/nn/bitserial_util.py]: "

**Comment by tqchen** [comment]
@cowanmeg please act on comments by @eqy then we can merge

**Comment by tqchen** [comment]
Thanks, @vinx13 @eqy @cowanmeg , this is now merged

**Pull request merged**